src/poetry/installation/chef.py

```diff
@@ -2,27 +2,206 @@
 
 import hashlib
 import json
+import tarfile
+import tempfile
+import threading
+import zipfile
 
+from contextlib import redirect_stdout
+from io import StringIO
 from pathlib import Path
 from typing import TYPE_CHECKING
+from typing import Callable
+from typing import Collection
+
+from build import BuildBackendException
+from build import ProjectBuilder
+from build.env import IsolatedEnv as BaseIsolatedEnv
+from pep517 import quiet_subprocess_runner  # type: ignore[import]
+from poetry.core.utils.helpers import temporary_directory
 
 from poetry.installation.chooser import InvalidWheelName
 from poetry.installation.chooser import Wheel
 
 
 if TYPE_CHECKING:
+    from contextlib import AbstractContextManager
+
     from poetry.core.packages.utils.link import Link
 
     from poetry.config.config import Config
     from poetry.utils.env import Env
 
 
+class ChefError(Exception):
+    ...
+
+
+class ChefBuildError(ChefError):
+    ...
+
+
+class IsolatedEnv(BaseIsolatedEnv):
+    def __init__(self, env: Env, config: Config) -> None:
+        self._env = env
+        self._config = config
+
+    @property
+    def executable(self) -> str:
+        return str(self._env.python)
+
+    @property
+    def scripts_dir(self) -> str:
+        return str(self._env._bin_dir)
+
+    def install(self, requirements: Collection[str]) -> None:
+        from cleo.io.null_io import NullIO
+        from poetry.core.packages.dependency import Dependency
+        from poetry.core.packages.project_package import ProjectPackage
+
+        from poetry.config.config import Config
+        from poetry.factory import Factory
+        from poetry.installation.installer import Installer
+        from poetry.packages.locker import NullLocker
+        from poetry.repositories.installed_repository import InstalledRepository
+
+        # We build Poetry dependencies from the requirements
+        package = ProjectPackage("__root__", "0.0.0")
+        package.python_versions = ".".join(str(v) for v in self._env.version_info[:3])
+        for requirement in requirements:
+            dependency = Dependency.create_from_pep_508(requirement)
+            package.add_dependency(dependency)
+
+        pool = Factory.create_pool(self._config)
+        installer = Installer(
+            NullIO(),
+            self._env,
+            package,
+            NullLocker(self._env.path.joinpath("poetry.lock"), {}),
+            pool,
+            Config.create(),
+            InstalledRepository.load(self._env),
+        )
+        installer.update(True)
+        installer.run()
+
+
 class Chef:
     def __init__(self, config: Config, env: Env) -> None:
+        self._config = config
         self._env = env
         self._cache_dir = (
             Path(config.get("cache-dir")).expanduser().joinpath("artifacts")
         )
+        self._lock = threading.Lock()
```

Comment:
@sdispater do we need locking here? In theory we already know all the `sdists` we need to build, why not just fan it out via multiprocess. Probably will have to build all required environments before hand if we want to cache/re-use though. But that should all be easily computed.

Reply:
naively removing this lock causes things to go wrong eg while it more than recovers the lost performance as described in https://github.com/python-poetry/poetry/pull/6205#issuecomment-1304814783, we also hit
```
 • Installing python-u2flib-server (5.0.1): Failed

  FileNotFoundError

  [Errno 2] No such file or directory: '/tmp/tmps4b9ofmc/petname-2.6'

  at ~/.virtualenvs/poetry/lib/python3.10/site-packages/build/__init__.py:149 in _working_directory
      145│
      146│     try:
      147│         yield
      148│     finally:
    → 149│         os.chdir(current)
      150│
      151│
      152│ def _validate_source_directory(srcdir: PathType) -> None:
      153│     if not os.path.isdir(srcdir):
```

https://github.com/python-poetry/poetry/pull/4213#issuecomment-1029786297 presumably?

Reply:
I think the reason is that `build` changes the working directory, which is not thread-safe: https://github.com/pypa/build/blob/63342ce0adc2333123aa63bf6b03c5d2235af324/src/build/__init__.py#L140-L149

With that knowledge we can use the lock more precisely which results in a significant performance improvement. Since this is a threading issue, multiprocessing should also be possible and might even be faster.

Reply:
build 0.10 now published, should include your fix at https://github.com/pypa/build/pull/536

can we remove this lock altogether now?

Reply:
Yes, I removed it.